buildSrc/src/main/java/org/elasticsearch/gradle/VersionCollection.java

```diff
@@ -0,0 +1,331 @@
+                new UnreleasedVersionDescription(
+                    version,
+                    getBranchFor(version),
+                    getGradleProjectNameFor(version)
+                )
+            ));
+    }
+
+    private String getGradleProjectNameFor(Version version) {
+        if (version.equals(currentVersion)) {
+            throw new IllegalArgumentException("The Gradle project to build " + version + " is the current build.");
+        }
+        Map<Integer, List<Version>> releasedMajorGroupedByMinor = getReleasedMajorGroupedByMinor();
+
+        if (version.getRevision() == 0) {
+            if (releasedMajorGroupedByMinor
+                .get(releasedMajorGroupedByMinor.keySet().stream().max(Integer::compareTo).orElse(0))
+                .contains(version)) {
+                return "minor";
+            } else {
+                return "staged";
+            }
+        } else {
+            if (releasedMajorGroupedByMinor
+                .getOrDefault(version.getMinor(), emptyList())
+                .contains(version)) {
+                return "maintenance";
+            } else {
+                return "bugfix";
+            }
+        }
+    }
+
+    private String getBranchFor(Version version) {
+        switch (getGradleProjectNameFor(version)) {
+            case "minor":
+                return version.getMajor() + ".x";
+            case "staged":
+            case "maintenance":
+            case "bugfix":
+                return version.getMajor() + "." + version.getMinor();
+            default:
+                throw new IllegalStateException("Unexpected Gradle project name");
+        }
+    }
+
+    public List<Version> getUnreleased() {
+        List<Version> unreleased = new ArrayList<>();
+        // The current version is being worked, is always unreleased
+        unreleased.add(currentVersion);
+
+        // the tip of the previous major is unreleased for sure, be it a minor or a bugfix
+        unreleased.add(getLatestVersionByKey(this.groupByMajor, currentVersion.getMajor() - 1));
+
+        final Map<Integer, List<Version>> groupByMinor = getReleasedMajorGroupedByMinor();
+        int greatestMinor = groupByMinor.keySet().stream().max(Integer::compareTo).orElse(0);
+
+        // the last bugfix for this minor series is always unreleased
+        unreleased.add(getLatestVersionByKey(groupByMinor, greatestMinor));
+
+        if (groupByMinor.get(greatestMinor).size() == 1) {
+            // we found an unreleased minor
+            unreleased.add(getLatestVersionByKey(groupByMinor, greatestMinor - 1));
+            if (groupByMinor.getOrDefault(greatestMinor - 1, emptyList()).size() == 1) {
+                // we found that the previous minor is staged but not yet released
+                // in this case, the minor before that has a bugfix
+                unreleased.add(getLatestVersionByKey(groupByMinor, greatestMinor - 2));
+            }
+        }
+
+        return unmodifiableList(
```

Comment:
We could construct this once in the ctor instead of each time this method is called?

Reply:
That feels to me a bit like premature optimization.
Doing it in the constructor and adding a field for it makes the class even harder to understand, I don't think it's worth it.